cpp/src/arrow/testing/fixed_width_test_util.h

```diff
@@ -0,0 +1,72 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/testing/visibility.h"
```

Comment:
We should try to "include what you use" [1] as much as possible. Even when headers that we need are already included by other headers.

- `<memory>`
- `<functional>`
- `<cstdint>`
- `"arrow/type.h"` for `DataType` (complete type is needed for the pointers and vector declarations)
- `"arrow/type_fwd.h"` forward-declares `ArrayBuilder` (which we don't need to be a complete type in the header because we only pointers/references to it in this header)

With these included, you should be able to remove `"arrow/builder.h"` which is a very expensive header.

[1] https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md